collector/prometheus.go

```diff
@@ -0,0 +1,224 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	costQuerier PrometheusConfig
+	promConn    prom.API
+
+	costMgmtNamespace   = "openshift-cost"
+	monitoringNamespace = "openshift-monitoring"
+	secretKey           = "token"
+	serviceAccountName  = "default"
+	thanosRouteName     = "thanos-querier"
+	tokenRegex          = "default-token-*"
+
+	certFile = "/var/run/configmaps/trusted-ca-bundle/ca-bundle.crt"
+)
+
+// PrometheusConfig provides the configuration options to set up a Prometheus connections from a URL.
+type PrometheusConfig struct {
+	// Address is the URL to reach Prometheus.
+	Address string
+	// BearerToken is the user auth token
+	BearerToken config.Secret
+	// CAFile is the ca file
+	CAFile string
+}
+
+func getRuntimeObj(ctx context.Context, r client.Client, obj runtime.Object, key types.NamespacedName, name string) error {
+	err := r.Get(ctx, key, obj)
+	if err != nil {
+		switch {
+		case errors.IsNotFound(err):
+			return fmt.Errorf("no %s found", name)
+		case errors.IsForbidden(err):
+			return fmt.Errorf("operator does not have permission to check %s", name)
+		default:
+			return fmt.Errorf("could not check %s: %v", name, err)
+		}
+	}
+	return nil
+}
+
+func getBearerToken(ctx context.Context, r client.Client, cfg *PrometheusConfig) error {
+	sa := &corev1.ServiceAccount{}
+	objKey := client.ObjectKey{
+		Namespace: costMgmtNamespace,
+		Name:      serviceAccountName,
+	}
+	err := getRuntimeObj(ctx, r, sa, objKey, "service account")
+	if err != nil {
+		return err
+	}
+
+	if len(sa.Secrets) <= 0 {
+		return fmt.Errorf("getBearerToken: no secrets in service account")
+	}
+
+	for _, secret := range sa.Secrets {
+		matched, _ := regexp.MatchString(tokenRegex, secret.Name)
+		if !matched {
+			continue
+		}
+
+		s := &corev1.Secret{}
+		objKey := client.ObjectKey{
+			Namespace: costMgmtNamespace,
+			Name:      secret.Name,
+		}
+		err := getRuntimeObj(ctx, r, s, objKey, "secret")
+		if err != nil {
+			return err
+		}
+		encodedSecret, ok := s.Data[secretKey]
+		if !ok {
+			return fmt.Errorf("getBearerToken: cannot find token in secret")
+		}
+		if len(encodedSecret) <= 0 {
+			return fmt.Errorf("getBearerToken: no data in default secret")
+		}
+		cfg.BearerToken = config.Secret(encodedSecret)
+		return nil
+	}
+	return fmt.Errorf("getBearerToken: no token found")
+
+}
+
+func getPromAddress(ctx context.Context, r client.Client, cfg *PrometheusConfig) error {
+	route := &routev1.Route{}
+	objKey := client.ObjectKey{
+		Namespace: monitoringNamespace,
+		Name:      thanosRouteName,
+	}
+	err := getRuntimeObj(ctx, r, route, objKey, "route")
+	if err != nil {
+		return err
+	}
+
+	if route.Spec.Host == "" {
+		return fmt.Errorf("getPromAddress: no routes found")
+	}
+	cfg.Address = "https://" + route.Spec.Host
+	return nil
```

Comment:
Is there a reason why we can't use the service directly instead of the route? If we are running internal to the cluster it doesn't seem necessary to have to use an external route. One of the recent Solution Architects actually called this out in our current operator for how we interact with the report operator and said that this breaks any customer that puts an SSO config in front of any external route.

Additionally, for potential work arounds it might make sense to have a SPEC config item to specify the prometheus address (for testing or other issues).

Reply:
I can certainly rework this one. I was having some trouble with the svc address (I can't remember why). But I just tried again with `https://thanos-querier.openshift-monitoring.svc:9091` in the cluster, and that worked just fine.

I'll work on a different solution for using an external route for local development.

Reply:
We don't have to block on this if you just want to do a follow up PR.

Reply:
Went ahead and took care of the route.
For local dev, we can add an external address to the CR under the `prometheus_config`:

```
spec:
  prometheus_config:
    address: >-
      https://thanos-querier-openshift-monitoring.apps.cluster....
status:
  authentication:
    credentials_found: true
    type: token
  clusterID: 6534a350-4f43-4393-9383-8e37939e9922
  ingress_url: 'https://cloud.redhat.com/api/ingress/v1/upload'
  operator_commit: f27bf0b8fa5f1aed9744b1b8d51e945fe7b63271
  prometheus:
    address: >-
      https://thanos-querier-openshift-monitoring.apps.cluster...
    last_query_start_time: '2020-10-14T22:39:00Z'
    last_query_success_time: '2020-10-14T22:39:00Z'
    prometheus_connected: true
```